### PR TITLE
API Custom admin URL

### PR DESCRIPTION
**Pull request by colymba**
Pull request that follows the short discussion on the mailing list with @chillu here https://groups.google.com/forum/#!topic/silverstripe-dev/gGVRd0wS0U8

This PR goes along with the CMS one here https://github.com/silverstripe/silverstripe-cms/pull/1041
There is probably a few modules that will need updating, not sure which ones yet?

So far still a bit of a work in progress until we are happy with the implementation.

In short, this PR does:
- Use `AdminRootController` as the main class for admin URL configuration etc.
- `AdminRootController.url_base` is the main/only config for custom URL
- This is used during bootstrap to create a `Director.rules` config entry via `AdminRootController::set_admin_route()`
- Remove all other instances of `url_base` configs in other classes
- Makes utility functions and variables available globally:
  - PHP `AdminRootController::adminURL()`
  - SS `$AdminURL`
  - JS `ss.config.adminURL`

To me this seemed like the best approach and made things easier by only having to configure 1 thing once, even though the bootstrap call might not be that elegant? What do you think?

Note: one thing left are `SiteTree.DEFAULTHOMECONTENT` translations that have the admin URL hard-coded, but this might not be an issue?

### CODE REVIEW

**Comment by patricknelson** [comment]
Just wanted to chime in here separately to bump the thread a bit and to flag this as something I'm interested in seeing as well. Let me know if you need any help with this. I'll say here too that I know "security through obscurity" isn't security at all. However, *adding* obscurity in conjunction with best practices shouldn't detract, so long as those practices aren't diminished in favor of such obscurity (i.e. not staying on top of security fixes).

Preaching to the choir here, I think :D

File: admin/code/AdminRootController.php [hunk 1]
```diff
@@ -4,15 +4,56 @@
...
+	 */
+	public static function get_admin_route() {
+		$override = Config::inst()->get('AdminRootController', 'url_base');
+
+		if ($override) {
+			return $override;
+		}
+		else{
+			$rules      = Config::inst()->get('Director', 'rules');
+			$adminRoute = array_search('AdminRootController', $rules);
```
Comment (chillu) [inline admin/code/AdminRootController.php]: Please don't align assignments. Even though its not specifically pointed out in the conventions, we should follow implicit conventions here - the vast majority of existing assignments don't align.

File: admin/code/AdminRootController.php [hunk 2]
```diff
@@ -4,15 +4,56 @@
...
+		}
+	}
+
+	/**
+	 * Returns the root admin URL for the site
+	 *
+	 * @return string
+	 * @uses get_admin_route()
+	 */
+	public static function adminURL() {
```
Comment (chillu) [inline admin/code/AdminRootController.php]: static methods should be snake case

File: docs/en/reference/cms-architecture.md [hunk 3]
```diff
@@ -15,6 +15,42 @@ with some special conventions.
...
+	$AdminURL
+
+And in JavaScript, this is avaible through the `ss` namespace
+
+	:::js
+	ss.config.adminURL
+
+### Multiple Admin URL and overrides
+
+You can also create your own classes that extend the `[api:AdminRootController]`, to create multiple or cutoms admin areas, with a `Director.rules` for each one. In those cases, you should set your classes' `url_base` static config to match your `Director.rules`, for the rest of the cms interface to work properly.
```
Comment (chillu) [inline docs/en/reference/cms-architecture.md]: Spelling mistake ("cutoms")

**Comment by chillu** [comment]
Looks mostly good (apart from the few niggly comments above), thanks for persisting on this! Can you please write a note in the upgrade guide pointing out changes if `admin/` has been hardcoded anywhere? Just an overview of changes, details can be linked off to `cms-architecture.md` (thanks for writing those btw!)

Also, could you please rebase?

**Comment by colymba** [comment]
Sorry for the delay... working on this now I'll update the PR asap...

**Comment by colymba** [comment]
Should be all up to date, squashed and ready to go, pending unit tests... Same with the CMS version: https://github.com/silverstripe/silverstripe-cms/pull/1041

File: control/Director.php [hunk 4]
```diff
@@ -351,12 +351,14 @@ protected static function handleRequest(SS_HTTPRequest $request, Session $sessio
 		if(isset($_REQUEST['debug'])) Debug::show($rules);
 
 		foreach($rules as $pattern => $controllerOptions) {
-			if(is_string($controllerOptions)) {
+			if(is_string($controllerOptions) && $controllerOptions) {
```
Comment (chillu) [inline control/Director.php]: Can you explain why this was necessary? Its changing the core routing, so want to ensure its solid

File: docs/en/04_Changelogs/3.2.0.md [hunk 5]
```diff
@@ -17,6 +17,7 @@
  * `Mailer` no longer calls `xml2raw` on all email subject line, and now must be passed in via plain text.
  * `ErrorControlChain` now supports reload on exceptions
  * `FormField::validate` now requires an instance of `Validator`
+ * Admin URL can now be configured via custom Director routing rule
```
Comment (chillu) [inline docs/en/04_Changelogs/3.2.0.md]: This is modifying the wrong changelog - your PR is against `master`, which will become the `4.0` release. Given this is a relatively low impact feature (no API change), I'd be happy for you to rebase this work against the `3` branch though (which will be released as 3.2 soon).

Comment (tractorcow) [inline docs/en/04_Changelogs/3.2.0.md]: It might need to be 4.0 if it breaks modules which rely on the admin url being fixed as 'admin'.

**Comment by tractorcow** [comment]
Do you think we should rebase this for 4.0?

**Comment by sminnee** [comment]
@tractorcow this would be useful functionality, so yeah it would be good to pull into `master` if we could...

**Comment by dhensby** [comment]
Bump

**Comment by tractorcow** [comment]
Who wants to update and rebase it for 4?

**Comment by tractorcow** [comment]
Ping. :)

**Comment by dhensby** [comment]
@tractorcow someone with time and an intricate knowledge of how all these files have morphed for 4?

**Comment by tractorcow** [comment]
I have only one of those things lol.

**Comment by tractorcow** [comment]
Fixes at https://github.com/silverstripe/silverstripe-framework/pull/5834

**Pull request merged**